luaui/Widgets/unit_smart_select.lua

```diff
@@ -207,39 +213,41 @@ function widget:Update()
 
 	local mouseSelection
 
-	if spIsAboveMiniMap(x1, y1) then
+	if inMiniMapSel then
 		mouseSelection = GetUnitsInMinimapRectangle(x1, y1, x2, y2)
 	else
 		mouseSelection = spGetUnitsInScreenRectangle(x1, y1, x2, y2, nil) or {}
 	end
 
-	inSelection = true
-
 	local newSelection = {}
 	local uid, udid, tmp
 
-	-- filter unselectable units
 	tmp = {}
+	local n = 0
+	local equalsMouseSelection = #mouseSelection == #lastMouseSelection
+	local isGodMode = spIsGodModeEnabled()
+
 	for i = 1, #mouseSelection do
 		uid = mouseSelection[i]
-		if not spGetUnitNoSelect(uid) then
-			tmp[#tmp + 1] = uid
+		if not spGetUnitNoSelect(uid) and -- filter unselectable units
+			(not isGodMode and spGetUnitTeam(uid) ~= GaiaTeamID and not ignoreUnits[spGetUnitDefID(uid)] and (spec or spGetUnitTeam(uid) == myTeamID)) then -- filter gaia units + ignored units (objects) + only own units when not spectating
```

Comment:
Here we join two loops to avoid 2n computations to n, additionally culls some entries for the last selection cache comparison.